packages/react-utilities/src/compose/types.ts

```diff
@@ -32,53 +34,86 @@ export type UnknownSlotProps = SlotProps<
 };
 
 /**
- * Defines the slot props for a slot that supports a Component type.
+ * The props type and shorthand value for a slot.
+ *
+ * The slot type can be either an intrinsic element like `'div'`, or a component like `typeof Button`.
  *
- * For intrinsic/native elements like 'div', use {@link IntrinsicSlotProps} instead.
+ * For intrinsic elements:
+ * * The first param is the slot's default type if no `as` prop is specified.
+ * * The second param is an optional union of alternative types that can be specified for the `as` prop.
  *
- * The generic param is the type of a control, i.e. a React component. For example:
+ * For components:
+ * * The first param is the component's type, like `typeof Button`
+ * * The second param is ignored and should not be specified
  *
  * @example
  * ```
- * type Props = {...}
- * const Button: React.FC<Props> = () => {...}
- * // $ExpectType ...
- * type SlotProps = ComponentSlotProps<typeof Button>
+ * // Intrinsic element examples:
+ * Slot<'div'> // Slot is always div
+ * Slot<'button', 'a'> // Defaults to button, but allows as="a" with anchor-specific props
+ * Slot<'label', 'span' | 'div'> // Defaults to label, but allows as="span" or as="div"
+ * NonNullable<Slot<'div'>> // Slot that will always be rendered and can't be set to null by the user
+ *
+ * // Component examples:
+ * Slot<typeof Button> // Slot is always a Button, and accepts all of Button's Props
+ * NonNullable<Slot<typeof Label>> // Slot is a Label and will always be rendered (can't be set to null by the user)
  * ```
  */
-export type ComponentSlotProps<Component extends React.ComponentType> = Component extends React.ComponentType<
-  infer Props
->
-  ? SlotProps<Props>
-  : never;
+export type Slot<
+  Type extends keyof JSX.IntrinsicElements | React.ComponentType,
+  AlternateAs extends keyof JSX.IntrinsicElements = never
+> =
+  | (Type extends keyof JSX.IntrinsicElements
+      ? IntrinsicSlotProps<Type, AlternateAs>
+      : Type extends React.ComponentType
+      ? ComponentSlotProps<Type>
+      : Type)
+  | SlotShorthandValue
+  | null;
 
 /**
- * Define the slot arguments for a slot that supports one or more intrinsic element types, such as 'div'.
- * For slots that support custom components, use {@link ComponentSlotProps} instead.
+ * Similar to {@link Slot}, but doesn't allow children to be specified through shorthand or props.
  *
- * The first param is the slot's default type if no `as` prop is specified.
- * The second param is an optional union of alternative types that can be specified for the `as` prop.
+ * This is intended to be used for element types that don't allow children, like `'input'` or `'img'`.
+ */
+export type SlotNoChildren<
+  Type extends keyof JSX.IntrinsicElements,
+  AlternateAs extends keyof JSX.IntrinsicElements = never
+> =
+  | (Omit<IntrinsicSlotProps<Type, AlternateAs>, 'children'> & {
+      children?: SlotRenderFunction<IntrinsicSlotProps<Type, AlternateAs>>;
+    })
+  | null;
```

Comment:
This will not work with custom components, correct? I mean `SlotNoChildren<typeof Input>`.

Reply:
Good point. At the time I wrote it, I didn't think that case was necessary because the component's `children` prop should already reflect that it doesn't allow children. However, I just realized that `Slot<typeof Input>` still allows shorthand even though Input doesn't allow children.

I think I'll update the types as follows:
* `Slot<>` should restrict the shorthand values to only allow things that are valid for the `children` prop of that slot. That would handle the `Slot<typeof Input>` case.
  * E.g. instead of Slot<> including `| SlotShorthandValue`, it would instead include `| Extract<SlotShorthandValue, Props['children']>`.
* Replace `SlotNoChildren<>` with the `WithoutChildren<>` helper so it can be composed into any slot. This is @spmonahan's suggestion: https://github.com/microsoft/fluentui/pull/21518#discussion_r797941059

Reply:
Actually based on https://github.com/microsoft/fluentui/pull/21518#discussion_r797406400, I'm going to remove `SlotNoChildren` entirely and just make `Slot` do the right thing automatically.